sql/core/src/main/java/org/apache/spark/sql/sources/v2/reader/SamplePushDownSupport.java

```diff
@@ -0,0 +1,32 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.sources.v2.reader;
+
+/**
+ * A mix in interface for `DataSourceV2Reader`. Users can implement this interface to push down
+ * "sample" operation to the data source and reduce the size of the data to be read.
+ */
+public interface SamplePushDownSupport {
+  /**
+   * Returns true if the implementation can handle this sample operation, so that we can reduce
+   * the data size to be read at the very beginning.
+   *
+   * TODO: shall we add more parameters like seed, lowerBound, upperBound, etc.?
```

Comment:
+1 for seed.

Reply:
seed is a must. i wouldn't do lowerbound / upperbound ... in virtually all cases people just do sample without replacement, so a fraction would be sufficient.

Reply:
we should also document the value range of "fraction".

Reply:
This should also document the assumptions that Spark has for a sample.

You can imagine a situation where I know the number of records ahead of time, so I turn `pushDownSample` into a `limit(fraction*numRecords)`, which wouldn't be correct. Maybe this should instead rely on reporting bucketing to the planner?